File: src/main/java/me/croco/lv2/Solution2_1.java
package me.croco.lv2;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * 240303
 * https://school.programmers.co.kr/learn/courses/30/lessons/12951
 *
 * JadenCase란 모든 단어의 첫 문자가 대문자이고, 그 외의 알파벳은 소문자인 문자열입니다. 단, 첫 문자가 알파벳이 아닐 때에는 이어지는 알파벳은 소문자로 쓰면 됩니다. (첫 번째 입출력 예 참고)
 * 문자열 s가 주어졌을 때, s를 JadenCase로 바꾼 문자열을 리턴하는 함수, solution을 완성해주세요.
 *
 * 제한 조건
 * s는 길이 1 이상 200 이하인 문자열입니다.
 * s는 알파벳과 숫자, 공백문자(" ")로 이루어져 있습니다.
 * 숫자는 단어의 첫 문자로만 나옵니다.
 * 숫자로만 이루어진 단어는 없습니다.
 * 공백문자가 연속해서 나올 수 있습니다.
 */
public class Solution2_1 {
    public static String solution(String s) {
        String answer = "";

        s = s.toLowerCase();

        List<String> list = new ArrayList<>();
        for(char tmp : s.toCharArray()) {
            list.add(String.valueOf(tmp));
        }

        for(int i = 0; i < list.size(); i++) {
            String a = s.substring(i, i+1);
            String b = i > 0 ? s.substring(i-1, i) : " ";

            if (b.equals(" ") && a.matches("^[a-z]$")) {
                list.set(i, a.toUpperCase());
            }

        };
        for(String c : list) {
            answer += c;
        }
        return answer;


        /*
        베스트

        String answer = "";
        String[] sp = s.toLowerCase().split("");
        boolean flag = true;

        for(String ss : sp) {
            answer += flag ? ss.toUpperCase() : ss;
            flag = ss.equals(" ") ? true : false;
        }

        return answer;
         */
    }



}

File: python/python/me/croco/lv2/Solution2_1.py
'''
240316
연습문제 > 행렬의 곱셈
https://school.programmers.co.kr/learn/courses/30/lessons/12949

문제 설명
2차원 행렬 arr1과 arr2를 입력받아, arr1에 arr2를 곱한 결과를 반환하는 함수, solution을 완성해주세요.

제한 조건
행렬 arr1, arr2의 행과 열의 길이는 2 이상 100 이하입니다.
행렬 arr1, arr2의 원소는 -10 이상 20 이하인 자연수입니다.
곱할 수 있는 배열만 주어집니다.

입출력 예
arr1	                              arr2	                                return
[[1, 4], [3, 2], [4, 1]]	            [[3, 3], [3, 3]]	                [[15, 15], [15, 15], [15, 15]]
[[2, 3, 2], [4, 2, 4], [3, 1, 4]]	 [[5, 4, 3], [2, 4, 1], [3, 1, 1]]	  [[22, 22, 11], [36, 28, 18], [29, 20, 14]]
'''

def solution(arr1, arr2) :
    n1, m1 = len(arr1), len(arr1[0])
    n2, m2 = len(arr2), len(arr2[0])

    result = [[0] * m2 for _ in range(n1)]

    for i in range(n1) :
        for j in range(m2) :

            for k in range(m1) :
                result[i][j] += arr1[i][k] * arr2[k][j]

    return result


'''
zip() 이라는 함수를 알게 됨!
zip(iterable1, iterable2, ...)

list = [1, 2]
list2 = ['a', 'b']
print(zip(list, list2))     # [(1, 'a'), (2, 'b')]
'''

# zip을 이용해 구현하기

def solution2(arr1, arr2):
    # arr2의 전치 행렬을 구함   * 전치행렬 : 기존 행렬의 행과 열을 바꾼 행렬
    arr2_transposed = list(zip(*arr2))

    result = [
        [
            sum(
                a*b for a, b in zip(row_a, col_b)
            ) for col_b in arr2_transposed
        ] for row_a in arr1
    ]

    return result






arr1 = [[2, 3, 2], [4, 2, 4], [3, 1, 4]]
arr2 = [[5, 4, 3], [2, 4, 1], [3, 1, 1]]
res = solution2(arr1, arr2)
print(res)
